presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp

```diff
@@ -887,23 +887,37 @@ VeloxQueryPlanConverter::toVeloxQueryPlan(
   const auto type = toLocalExchangeType(node->type);
 
   const auto outputType = toRowType(node->partitioningScheme.outputLayout);
-  // Input layout from source, describing how data should be fed to our node.
-  // For all sources the layout should be the same, so we store only one (we
-  // use the 1st source for that).
-  // This layout and the output layout for the 1st source would be used to
-  // created the column mapping in the operator.
-  RowTypePtr inputTypeFromSource = toRowType(node->inputs[0]);
+
+  // Different source nodes may have different output layouts.
+  // Add ProjectNode on top of each source node to re-arrange the output columns
+  // to match the output layout of the LocalExchangeNode.
+  for (auto i = 0; i < sourceNodes.size(); ++i) {
+    auto names = outputType->names();
+    std::vector<std::shared_ptr<const ITypedExpr>> projections;
+    projections.reserve(outputType->size());
+
+    auto desiredSourceOutput = toRowType(node->inputs[i]);
+
+    for (auto j = 0; j < outputType->size(); j++) {
+      projections.emplace_back(std::make_shared<core::FieldAccessTypedExpr>(
+          outputType->childAt(j), desiredSourceOutput->nameOf(j)));
+    }
+
+    sourceNodes[i] = std::make_shared<velox::core::ProjectNode>(
+        fmt::format("{}.{}", node->id, i),
+        std::move(names),
+        std::move(projections),
+        sourceNodes[i]);
+  }
```

Comment:
So, this will also take care of different input names?

Reply:
Yes. The extra project nodes take care of putting columns in the right order using the right names.